### PR TITLE
Change courses API to 2.4

### PR DESCRIPTION
**Pull request by Karthik-0**
### Changes done
- Changed courses API to 2.4

### CODE REVIEW

File: core/src/main/java/in/testpress/models/greendao/Chapter.java [hunk 1]
```diff
@@ -277,6 +277,45 @@ public void setActive(Boolean active) {
         }
         return getCourseChaptersQueryBuilder(context, courseId).where(parentCondition);
     }
+
+    public Integer getChildrenCountFromDB(Context context) {
```
Comment (bharathwaaj) [inline core/src/main/java/in/testpress/models/greendao/Chapter.java]: Should be hasChildren

File: core/src/main/java/in/testpress/models/greendao/Chapter.java [hunk 2]
```diff
@@ -277,6 +277,45 @@ public void setActive(Boolean active) {
...
+        List<Chapter> chaptersFromDB = chapterDao.queryBuilder()
+                .where(
+                        ChapterDao.Properties.ParentId.eq(getId()),
+                        ChapterDao.Properties.Active.eq(true)
+                ).list();
+
+        return chaptersFromDB.size();
+    }
+
+    public Integer getContentsCountFromDB(Context context) {
```
Comment (bharathwaaj) [inline core/src/main/java/in/testpress/models/greendao/Chapter.java]: Should be hasContents

File: core/src/main/java/in/testpress/models/greendao/Chapter.java [hunk 3]
```diff
@@ -277,6 +277,45 @@ public void setActive(Boolean active) {
...
+        List<Content> contentsFromDB = contentDao.queryBuilder()
+                .where(
+                        ContentDao.Properties.ChapterId.eq(getId()),
+                        ContentDao.Properties.Active.eq(true)
+                ).list();
+
+        return contentsFromDB.size();
+    }
+
+    public static Chapter getChapter(Context context, long chapterId) {
```
Comment (bharathwaaj) [inline core/src/main/java/in/testpress/models/greendao/Chapter.java]:
```suggestion
    public static Chapter get(Context context, long chapterId) {
```

File: core/src/main/java/in/testpress/models/greendao/Content.java [hunk 4]
```diff
@@ -693,6 +695,12 @@ public String getName() {
     public HtmlContent getHtml() {
         return getHtmlContent();
     }
+
+    public static QueryBuilder<Content> getCourseContents(Context context, Integer courseId) {
```
Comment (bharathwaaj) [inline core/src/main/java/in/testpress/models/greendao/Content.java]: Should go inside Course

File: course/src/main/java/in/testpress/course/ui/ChapterDetailActivity.java [hunk 5]
```diff
@@ -178,11 +178,11 @@ void onChapterLoaded(Chapter chapter) {
         this.chapter = chapter;
         //noinspection ConstantConditions
         getSupportActionBar().setTitle(chapter.getName());
-        if (chapter.getActive() && chapter.getChildrenCount() > 0) {
+        if (chapter.getActive() && chapter.getChildrenCountFromDB(this) > 0) {
```
Comment (bharathwaaj) [inline course/src/main/java/in/testpress/course/ui/ChapterDetailActivity.java]: Use hasChildren

File: course/src/main/java/in/testpress/course/ui/ChaptersListFragment.java [hunk 6]
```diff
@@ -99,12 +87,20 @@ private void setLastModifiedForPager() {
...
             }
         }
         return pager;
     }
 
+    @Override
+    public void onLoadFinished(Loader<List<Chapter>> loader,
+                               List<Chapter> items) {
+        super.onLoadFinished(loader, items);
+        Chapter.getParentChaptersQueryBuilder(getContext(), courseId, parentId).buildDelete().executeDeleteWithoutDetachingEntities();
```
Comment (bharathwaaj) [inline course/src/main/java/in/testpress/course/ui/ChaptersListFragment.java]: Remove getParentChaptersQueryBuilder and use chapter.children

Comment (bharathwaaj) [reply course/src/main/java/in/testpress/course/ui/ChaptersListFragment.java]: Also remove getCourseChaptersQueryBuilder and use course.chapters

File: course/src/main/java/in/testpress/course/ui/ContentActivity.java [hunk 7]
```diff
@@ -346,51 +354,80 @@ protected boolean shouldOverrideUrlLoading(Activity activity, String url) {
...
+
+        if (!chapters.isEmpty()) {
+            //noinspection ConstantConditions
+            getSupportActionBar().setTitle(chapters.get(0).getName());
+        }
+    }
+
     private void checkContentType() {
+        setChapterNameInActionBarTitle(content.getChapterId());
+        setContentTitle(content.getTitle());
```
Comment (bharathwaaj) [inline course/src/main/java/in/testpress/course/ui/ContentActivity.java]: Should be setTitle

File: course/src/main/java/in/testpress/course/ui/ContentActivity.java [hunk 8]
```diff
@@ -346,51 +354,80 @@ protected boolean shouldOverrideUrlLoading(Activity activity, String url) {
         }
     }
 
+    private void setChapterNameInActionBarTitle(long chapterId) {
+        List<Chapter> chapters = TestpressSDKDatabase.getChapterDao(this).queryBuilder()
+                .where(ChapterDao.Properties.Id.eq(chapterId)).list();
+
+        if (!chapters.isEmpty()) {
```
Comment (bharathwaaj) [inline course/src/main/java/in/testpress/course/ui/ContentActivity.java]: Remove this check

File: course/src/main/java/in/testpress/course/ui/ContentActivity.java [hunk 9]
```diff
@@ -346,51 +354,80 @@ protected boolean shouldOverrideUrlLoading(Activity activity, String url) {
...
+                initExoPlayer(video.getHlsUrl());
             }
-        } else if (content.getRawExam() != null) {
-            onExamContent();
-        } else if (content.getRawAttachment() != null) {
-            displayAttachmentContent();
-        } else {
-            setEmptyText(R.string.testpress_error_loading_contents,
-                    R.string.testpress_some_thing_went_wrong_try_again,
-                    R.drawable.ic_error_outline_black_18dp);
```
Comment (bharathwaaj) [inline course/src/main/java/in/testpress/course/ui/ContentActivity.java]: Undo all the changes not related to this PR

File: course/src/main/java/in/testpress/course/ui/ContentActivity.java [hunk 10]
```diff
@@ -346,51 +354,80 @@ protected boolean shouldOverrideUrlLoading(Activity activity, String url) {
...
+        List<Chapter> chapters = TestpressSDKDatabase.getChapterDao(this).queryBuilder()
+                .where(ChapterDao.Properties.Id.eq(chapterId)).list();
+
+        if (!chapters.isEmpty()) {
+            //noinspection ConstantConditions
+            getSupportActionBar().setTitle(chapters.get(0).getName());
+        }
+    }
+
     private void checkContentType() {
```
Comment (bharathwaaj) [inline course/src/main/java/in/testpress/course/ui/ContentActivity.java]: checkContentType is doing one thing but name is telling another thing

File: course/src/main/java/in/testpress/course/ui/ContentActivity.java (hunk 7 above)
Comment (Karthik-0) [reply course/src/main/java/in/testpress/course/ui/ContentActivity.java]: setTitle is android's default method to change activity's title

**Pull request merged**